Add test for trace

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47126 Port CPU Trace from TH to ATen
* **#47125 Add test for trace**

We didn't actually have any tests for torch.trace. The tests expose a
discrepancy between the behavior of torch.trace on CPU and CUDA that
I'll file an issue for.

Differential Revision: [D24683260](https://our.internmc.facebook.com/intern/diff/D24683260)

zou3519 committed:
Add test for trace

We didn't actually have any tests for torch.trace. The tests expose a
discrepancy between the behavior of torch.trace on CPU and CUDA that
I'll file an issue for.

[ghstack-poisoned]

zou3519 committed:
Update on "Add test for trace"


We didn't actually have any tests for torch.trace. The tests expose a
discrepancy between the behavior of torch.trace on CPU and CUDA that
I'll file an issue for.

[ghstack-poisoned]

anjali411 commented on test/test_torch.py:
```diff
@@ -5993,6 +5993,37 @@ def test_diagonal_multidim(self, device, dtype):
+            tensor = make_tensor(shape, device, dtype, low=-9, high=9)
+            diag = tensor.diag()
+            if legacy:
+                # NB: trace on cpu doesn't do type promotion...
```
good catch! once the issue is filed, can you also link it here in the comment?

zou3519 replied:
https://github.com/pytorch/pytorch/issues/47127

zou3519 replied:
tossed into comment

anjali411 commented on test/test_torch.py:
```diff
@@ -5993,6 +5993,37 @@ def test_diagonal_multidim(self, device, dtype):
+                expected = tensor.diag().sum().to(dtype)
+            else:
+                expected = tensor.diag().sum()
+            self.assertEqual(tensor.trace(), expected)
```
I think it would be nice to compare it to `np.trace()` result here

zou3519 replied:
That feels redundant, especially since numpy has slightly different type promotion semantics so we'll end up trying to force the result to the correct dtype

anjali411 replied:
isn't our type promotion different from that of numpy only while dealing with `torch.half` dtype? CCing @mruberry since he is working on testing and numpy compatibility

mruberry replied:
Comparing with NumPy and citing the issue would still be preferable. NumPy's trace supports a dtype argument that should make writing the comparison easy.

zou3519 replied:
Is there a torch-to-numpy dtype conversion somewhere?

mruberry replied:
This dict and its sister dict (see below it) handle the conversion for every type except bfloat16: 

https://github.com/pytorch/pytorch/blob/c10aa44e337d27b46efc997a16391d927bac8500/torch/testing/_internal/common_utils.py#L359

For bfloat16 you can cast to float32 when running on NumPy.

zou3519 replied:
Thanks!

anjali411 reviewed (approved):
left minor comments but looks good overall

zou3519 committed:
Update on "Add test for trace"


We didn't actually have any tests for torch.trace. The tests expose a
discrepancy between the behavior of torch.trace on CPU and CUDA that
I'll file an issue for.

[ghstack-poisoned]

zou3519 committed:
Update on "Add test for trace"


We didn't actually have any tests for torch.trace. The tests expose a
discrepancy between the behavior of torch.trace on CPU and CUDA that
I'll file an issue for.

Differential Revision: [D24683260](https://our.internmc.facebook.com/intern/diff/D24683260)

[ghstack-poisoned]
